query-engine/core/src/query_graph_builder/write/utils.rs

```diff
@@ -218,16 +207,19 @@ pub fn insert_1to1_idempotent_connect_checks(
     graph.create_edge(
         &diff_node,
         &if_node,
-        QueryGraphDependency::DataDependency(Box::new(move |if_node, result| {
-            let diff_result = result.as_diff_result().unwrap();
-            let should_connect = !diff_result.is_empty();
```

Comment:
Before this would check that any of the two diffs is not empty. Translating this literally would require introducing a third node type (`XorRecords` or something) but it's enough to check only `left - right` because both can contain at most one record (since it's 1:1 relationship) and we already checked that `left` contains exactly one record above.